network/hive.go

```diff
@@ -131,6 +137,121 @@ func (h *Hive) Stop() error {
+// Register interface method
+func (h *Hive) handlePeersMsg(_ context.Context, p *Peer, msg *peersMsg) error {
+	// register all addresses
+	if len(msg.Peers) == 0 {
+		return nil
+	}
+	for _, a := range msg.Peers {
+		p.seen(a)
+		h.BroadcastPeer(a)
+	}
+	return h.Kademlia.Register(msg.Peers...)
+}
+
+// subPeers msg is communicating the depth of the overlay table of a peer
+type subPeersMsg struct {
+	Depth uint8
+}
+
+// String returns the pretty printer
+func (msg subPeersMsg) String() string {
+	return fmt.Sprintf("%T: request peers > PO%02d. ", msg, msg.Depth)
+}
+
+// handleSubPeersMsg handles incoming subPeersMsg
+// this message represents the saturation depth of the remote peer
+// saturation depth is the radius within which the peer subscribes to peers
+// the first time this is received we send peer info on all
+// our connected peers that fall within peers saturation depth
+// otherwise this depth is just recorded on the peer, so that
+// subsequent new connections are sent iff they fall within the radius
+func (h *Hive) handleSubPeersMsg(ctx context.Context, p *Peer, msg *subPeersMsg) error {
+	p.setDepth(msg.Depth)
+	// only send peers after the initial subPeersMsg
+	if !p.sentPeers {
+		var peers []*BzzAddr
+		// iterate connection in ascending order of disctance from the remote address
+		h.EachConn(p.Over(), 255, func(kp *Peer, po int) bool {
+			// terminate if we are beyond the radius
+			if uint8(po) < msg.Depth {
+				return false
+			}
+			if !p.seen(kp.BzzAddr) { // here just records the peer sent
+				peers = append(peers, kp.BzzAddr)
+			}
+			return true
+		})
+		// if useful  peers are found, send them over
+		if len(peers) > 0 {
+			go p.Send(ctx, &peersMsg{Peers: sortPeers(peers)})
+		}
+
+		// on bootnode mode, kick the peer out after N records sent
+		if h.BootnodeMode && len(p.peers) > h.BootnodeKickAfterNPeers {
+			p.Disconnect(p2p.DiscRequested)
+			return nil
+		}
+	}
+	p.sentPeers = true
+	return nil
+}
+
+// BroadcastPeer informs all peers about a newly added node
+func (h *Hive) BroadcastPeer(p *BzzAddr) {
+	f := func(val *Peer, po int) bool {
+		h.AdvertisePeer(val, p, uint8(po))
+		return true
+	}
+	h.Kademlia.EachConn(p.Address(), 255, f)
+}
+
+// NotifyPeer notifies the remote node (recipient) about a peer if
+// the peer's PO is within the recipients advertised depth
+// OR the peer is closer to the recipient than self
+// unless already notified during the connection session
+func (h *Hive) AdvertisePeer(p *Peer, a *BzzAddr, po uint8) {
+	// immediately return
+	if (po < p.getDepth() && pot.ProxCmp(h.Kademlia.BaseAddr(), p.Address(), a.Address()) != 1) || p.seen(a) {
+		return
+	}
+	resp := &peersMsg{
+		Peers: []*BzzAddr{a},
+	}
+	log.Warn("notifypeer", "notify", resp)
+	p.Send(context.Background(), resp)
+}
+
+// NotifyDepth sends a message to all connections if depth of saturation is changed
+func (h *Hive) BroadcastDepth(depth uint8, kad *Kademlia) {
```

Comment:
No need to pass kademlia as an argument anymore. And comment has the old method name.